src/components/MainSidebar/DimensionsList/DimensionsList.js

```diff
@@ -0,0 +1,111 @@
+import i18n from '@dhis2/d2-i18n'
+import { CircularLoader, NoticeBox } from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React, { useRef, useEffect } from 'react'
+import { DimensionListItem } from './DimensionListItem.js'
+import styles from './DimensionsList.module.css'
+
+const getNoResultsMessage = (searchTerm, programName) => {
+    if (searchTerm) {
+        return i18n.t("No dimensions found for '{{searchTerm}}'", {
+            searchTerm,
+        })
+    }
+    if (programName) {
+        return i18n.t('No dimensions found in the {{programName}} program', {
+            programName,
+        })
+    }
+    return i18n.t('No dimensions found')
+}
+
+const DimensionsList = ({
+    loading,
+    fetching,
+    error,
+    dimensions,
+    programName,
+    searchTerm,
+    setIsListEndVisible,
+}) => {
+    const scrollBoxRef = useRef()
+    const updateIsListEndVisible = () => {
+        const el = scrollBoxRef.current
+        const isVisible = el.scrollHeight - el.scrollTop - el.clientHeight < 5
+
+        setIsListEndVisible(isVisible)
+    }
+
+    useEffect(() => {
+        if (dimensions && scrollBoxRef.current && !loading && !fetching) {
+            updateIsListEndVisible()
+        }
+    }, [dimensions, scrollBoxRef.current])
+
+    if (loading) {
+        return (
+            <div className={styles.loaderWrap}>
+                <CircularLoader small />
+            </div>
+        )
+    }
+
+    if (error) {
+        return (
+            <div className={styles.errorWrap}>
+                <NoticeBox error title={i18n.t("Couldn't load dimensions")}>
+                    {error?.message ||
+                        i18n.t(
+                            'There was a problem loading the dimensions. Try again, or contact your system administrator.'
+                        )}
+                </NoticeBox>
+            </div>
+        )
+    }
+
+    if (dimensions.length === 0) {
```

Comment:
```suggestion
    if (!dimensions?.length) {
```
`dimensions` isn't `isRequired` and has no `defaultProps` value (speaking of which, this in itself is a lint warning I think?), could thus be `undefined`?

Reply:
Once loading is done, and there is no error, then we have a dimensions array. I think rendering a "no results found message" when there is a dimension array which is empty, expresses the intent of what this block of code does very well.

The suggested change makes the code more robust, but I would prefer it to fail fast.

Hope you are OK with keeping this as-is.

Reply:
Ok, if you want it to fail if it (for whatever reason) doesn't have a `dimension` at all then maybe that warrants a `try catch` so we can provide a user-friendly error? 
I mean, the argument "it doesn't fail now so we don't need it to be more robust" only holds up for so long, right, so the undefined-check was just a suggestion to provide some future-proofing (+ personally I think checking `!array.length` is just as expressive as `array.length === 0` so I think your intent is still very clear).